### DISCUSSION TITLE
multi requiredMod in single file - combine equip of different dlc/mods

### DISCUSSION THREAD

**Discussion by sashaman** [Royal-Modding/FS19_BetterContracts#16] [Ideas]
as title says, 
else its practically impossible to combine various dlc equip ... see below example where the arion is not possible
would be fitting in vehicle key maybe ? 
so instead of:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<missionVehicles requiredMod="FS19_claasDominator108SL">

    <mission type="harvest">
        <!-- grainMission harvest -->
        <group fieldSize="small" variant="GRAIN">
            <vehicle filename="vehicles/claas/dominator108SL/dominator108SL.xml" > <!-- combine -->
                <configuration name="buyableGPS" id="2" />
				<configuration name="universalPassenger" id="1"/>
            </vehicle>
            <vehicle filename="vehicles/claas/c510/c510.xml" /> <!-- header -->
            <vehicle filename="vehicles/claas/c510Trailer/c510Trailer.xml" /> <!-- headerTrailer -->
            <vehicle filename="$pdlc/claasPack/vehicles/claas/arion400/arion400.xml" > <!-- tractor -->
                <configuration name="motor" id="4" />
                <configuration name="wheel" id="6" />
				<configuration name="tireSound" id="6"/>
				<configuration name="soundExtension" id="2"/>
            </vehicle>
            <vehicle filename="$data/vehicles/strautmann/SEK802/SEK802.xml" > <!-- trailer -->
                <configuration name="fillVolume" id="4" />
                <configuration name="trailer" id="3" />
                <configuration name="fillUnit" id="4" />
            </vehicle>
        </group>
    </mission>
</missionVehicles>
```
maybe like that:
```xml
<?xml version="1.0" encoding="UTF-8"?>
    <mission type="harvest">
        <!-- grainMission harvest -->
        <group fieldSize="small" variant="GRAIN">
            <vehicle filename="vehicles/claas/dominator108SL/dominator108SL.xml" requiredMod="FS19_claasDominator108SL"> <!-- combine -->
                <configuration name="buyableGPS" id="2" />
				<configuration name="universalPassenger" id="1"/>
            </vehicle>
            <vehicle filename="vehicles/claas/c510/c510.xml" requiredMod="FS19_claasDominator108SL"/> <!-- header -->
            <vehicle filename="vehicles/claas/c510Trailer/c510Trailer.xml" requiredMod="FS19_claasDominator108SL"/> <!-- headerTrailer -->
            <vehicle filename="$pdlc/claasPack/vehicles/claas/arion400/arion400.xml" requiredMod="pdlc_claasPack"> <!-- tractor -->
                <configuration name="motor" id="4" />
                <configuration name="wheel" id="6" />
				<configuration name="tireSound" id="6"/>
				<configuration name="soundExtension" id="2"/>
            </vehicle>
            <vehicle filename="$data/vehicles/strautmann/SEK802/SEK802.xml" > <!-- trailer -->
                <configuration name="fillVolume" id="4" />
                <configuration name="trailer" id="3" />
                <configuration name="fillUnit" id="4" />
            </vehicle>
        </group>
    </mission>
</missionVehicles>
```

**Comment by sashaman** [comment]
p.s. i dont know how difficult that would make a contract compilation, in regards to the different avail of used dlc/mods

**Reply by TyKonKet** [reply]
The idea is very good, we discussed something similar internally in the past, the biggest issue is exactly what you said, players have different mods/dlcs available / activated in their savegames.

**Comment by TyKonKet** [comment]
Another issue I see there is:
``` xml
<configuration name="buyableGPS" id="2" />
<configuration name="universalPassenger" id="1"/>
```

If a player hasn't one of those mods, it will lead to errors and to a not working contract system.

**Comment by Lactic68** [comment]
Maybe the solution would be to specify which mod the user is using by selecting correct checkboxes (like in Globalcompany).
Anyway this would quite complexify dev and maintenance of your mod I think.

**Reply by TyKonKet** [reply]
Yes, both your cons are right but there's something even more problematic, vehicle combos for missions must be generated at startup time, before you start playing.

**Comment by sashaman** [comment]
how about a (user-) configurable list
like having the dlc checks include a txt file where (the advanced? lol) users can tweak and specific mods can be added to or removed from
gps + UP are leftovers from my own edit lol but quite a good example of mods not too few users might use

**Comment by ghost** [comment]
I think I could try to implement s.th. from @sashaman's suggestion: when parsing the missionVehicles.xml 
```
<group fieldSize="small" variant="GRAIN">
    <vehicle filename="vehicles/claas/dominator108SL/dominator108SL.xml" requiredMod="FS19_claasDominator108SL"> <!-- combine -->
        <configuration name="buyableGPS" id="2" requiredMod="FS19_guidanceSteering" />
	<configuration name="universalPassenger" id="1" requiredMod="FS19_UniversalPassenger"/>
     </vehicle>
     <vehicle filename="vehicles/claas/c510/c510.xml" requiredMod="FS19_claasDominator108SL"/> <!-- header -->
     <vehicle filename="vehicles/claas/c510Trailer/c510Trailer.xml" requiredMod="FS19_claasDominator108SL"/> <!-- headerTrailer -->
```
the code would check if the indicated mods are loaded. If not, it would just skip this vehicle (or just the configuration). That could lead to funny vehicle combos offered for an ingame mission, e.g. a plow, a weight, but no tractor for a plowing mission. Or to an empty vehicle group, or even an empty group list for a specific mission type / fieldsize. But (in response to #26) BetterContracts now can also handle missions with no vehicle/equipment combo offered. And players could then still take such a contract, they'd just have to use own equipment.

**Comment by ghost** [comment]
Ok done. I just committed the extension for optional user-defined mission vehicles to main. You can test the [pre-Release](https://github.com/Royal-Modding/FS19_BetterContracts/releases/download/1.2.0.0-RC4/FS19_BetterContracts.zip).
@sashaman, you gave me a hard time during testing, when I tried to parse
```xml
  <vehicle filename="$pdlc/claasPack/vehicles/claas/arion400/arion400.xml" requiredMod="pdlc_claasPack"> <!-- tractor -->
        <configuration name="motor" id="4" />
        <configuration name="wheel" id="6" />
	<configuration name="tireSound" id="6"/>
	<configuration name="soundExtension" id="2"/
  </vehicle>
```
You specify `requiredMod="pdlc_claasPack"` ok, but where do we find "tireSound", "soundExtension"? The code now will ignore any mod configurations unknown to g_missionManager. For better readability, you can/ should now include 
```xml
  <vehicle filename="$pdlc/claasPack/vehicles/claas/arion400/arion400.xml" requiredMod="pdlc_claasPack"> <!-- tractor -->
      ...
	<configuration name="soundExtension" id="2" requiredMod="FS19_changedMotorSounds"/>
  </vehicle>
```
PS. See also #26 😉

**Reply by sashaman** [reply]
indeed, sry for any confusion caused. 
iirc at the time writing i could only specifiy one requiredMod thats the reason for the thread. 
honestly i didnt always have my additions in mind :P

**Reply by sashaman** [reply]
p.s. 
configuration name="tireSound" -> FS19_tireSound
configuration name="soundExtension"  -> FS19_changedMotorSounds
